src/main/java/org/truffleruby/builtins/CoreMethodNodeManager.java

```diff
@@ -304,7 +303,11 @@ public RubyNode createCoreMethodNode(NodeFactory<? extends RubyNode> nodeFactory
         }
 
         if (method.needsBlock()) {
-            argumentsNodes[i++] = new ReadBlockFromCurrentFrameArgumentsNode(NotProvided.INSTANCE);
+            /* The way we write specializations for getting a block or not is that we use NotProvided like a missing
+             * argument. The block coming into the method is actually always Nil or RubyProc, so here we check which and
+             * convert Nil to NotProvided. */
+            argumentsNodes[i++] = new ReadBlockFromCurrentFrameArgumentsNode.ConvertNilBlockToNotProvidedNode(
+                    new ReadBlockFromCurrentFrameArgumentsNode());
```

Comment:
I'd much rather we adapt the specialziation to use `Nil` instead of `NotProvided` for the block argument.
I think we should adapt [annotation processor checks](https://github.com/oracle/truffleruby/blob/master/src/processor/java/org/truffleruby/processor/CoreModuleChecks.java) to check that for `@CoreMethod(needsBlock = true)` then the argument could only be typed as RubyProc, Nil, or `Object maybeBlock`.

Reply:
> I'd much rather we adapt the specialziation to use Nil instead of NotProvided for the block argument.

We just didn't do that in this PR for churn reasons. That, and possibly also simplifying some places that pack arguments, can be simplified after this PR.

Reply:
I'd be fine to have the rest of the changes in another PR, but I think we don't want to keep the half-migration like here too long. It also seems a low-hanging fruit to finish the migration with this PR in.

Reply:
Actually this change I get we might want to do it separately, and it touches a fairly different part of the code.
I'd do https://github.com/oracle/truffleruby/pull/2232#discussion_r562627254 in this PR though.